File: ios/TapScore/Features/Create/FormationCatalog.swift
import Foundation

/// The shared-ball formations the server serves — scramble, foursomes,
/// greensomes — and the seeding recipe each one carries.
///
/// The Swift image of `server/domain/round-setup/formation-catalog.ts`, read
/// over `GET /setup/formations` exactly as `FormatCatalog` reads
/// `GET /setup/formats`. Like it, this is a **value over the server's
/// descriptors**, never a local table: an allowance recipe that disagrees with
/// the server's is a scorecard nobody can explain, and there is no way to spot
/// the disagreement from the phone.
///
/// So there is deliberately **no hardcoded fallback**. A catalog that failed to
/// load is simply EMPTY, `isAvailable` is false, and shared-ball teams cannot be
/// built in that session — the flow is otherwise untouched (proposal
/// `docs/proposals/ball-teams-composition.md`, "Formations are rule objects,
/// not metadata").
///
/// `custom` is absent by design, server-side and here: it is a valid draft
/// `formation` value with no recipe, reachable only from the web flexible
/// editor. A stored team carrying it round-trips through this client as
/// read-only passthrough (`EditDraftAssembler`), never as an editable team.
struct FormationCatalog: Sendable, Equatable {
    var descriptors: [FormationDescriptor] = []

    init(descriptors: [FormationDescriptor] = []) {
        self.descriptors = descriptors
    }

    /// The feature is available at all — i.e. the catalog fetch succeeded.
    var isAvailable: Bool { !descriptors.isEmpty }

    // MARK: - Lookup

    func byId(_ id: String) -> FormationDescriptor? {
        descriptors.first { $0.id == id }
    }

    /// `labels[locale] ?? labels.en` — the same "never throw on an unknown id"
    /// contract `FormatCatalog.label` has, so the caller keeps its own fallback.
    func label(_ id: String, locale: FormatCatalog.Locale = .current) -> String? {
        byId(id).map { label($0, locale: locale) }
    }

    func label(_ descriptor: FormationDescriptor, locale: FormatCatalog.Locale = .current) -> String {
        switch locale {
        case .sv:
            if let sv = descriptor.labels.sv, !sv.isEmpty { return sv }
            return descriptor.labels.en
        case .en:
            return descriptor.labels.en
        }
    }

    // MARK: - Rules

    /// The team-size bounds, inclusive. Nil for a formation this catalog does
    /// not know — which is the one state in which the caller must not enforce
    /// anything, rather than enforce a guess.
    func size(_ id: String) -> (min: Int, max: Int)? {
        byId(id).map { (min: Int($0.size.min), max: Int($0.size.max)) }
    }

    func fits(_ id: String, memberCount: Int) -> Bool {
        guard let size = size(id) else { return true }
        return memberCount >= size.min && memberCount <= size.max
    }

    /// The default allowance % **by position**, with members sorted by playing
    /// handicap ASCENDING (position 1 = the lowest handicap). Nil when this
    /// formation has no recipe for that many members — a size the bounds
    /// already refuse, or a formation the catalog does not carry.
    ///
    /// The wire spells the key as a string because that is what a JSON object
    /// carries; `allowancesBySize["2"]` is the pair recipe.
    func allowances(_ id: String, memberCount: Int) -> [Double]? {
        guard let recipe = byId(id)?.allowancesBySize[String(memberCount)],
              recipe.count == memberCount
        else { return nil }
        return recipe
    }
}

File: server/domain/round-setup/formation-catalog.ts
// Ball teams (docs/proposals/ball-teams-composition.md) Phase A — the
// serializable formation catalog.
//
// A `formation` on a `single_ball` `DraftRoundTeam` (draft.ts) used to be pure
// metadata: a display label the builder read only to name the merged ball. It
// becomes a RULE OBJECT here — size bounds plus the default per-position
// allowance recipe the setup UI seeds member percentages from. The draft schema
// is UNCHANGED: `DraftRoundTeam` already carries `formation` + per-member
// `allowancePct`, and the builder still never needs the recipe. Seeding is a
// setup-time concern; what reaches the builder is always frozen percentages.
//
// This is a fixed table, not a plugin registry — a formation carries no
// behaviour to dispatch (contrast `../formats/plugin.ts` and
// `../aggregation/strategy.ts`, which register functions). So: no registry, no
// `register*` seam, just the descriptors, served by `GET /setup/formations`.
// Both clients read this one source; adding a formation never touches client
// code.
//
// `custom` is deliberately ABSENT: it stays a valid draft `formation` value for
// the web flexible editor (unbounded size, hand-entered percentages) precisely
// because it has no recipe to describe.
//
// The server does NOT enforce these bounds — single-ball team size is
// unvalidated server-side today (`DraftTeamMember` has `minItems: 1` and no
// max; builder.ts checks kind/nesting only). Bounds are a setup-UI concern
// like the rest of the seeding, alongside the web client's `MAX_TEAM_SIZE`
// (src/create/format-catalog.service.ts) = 10.
//
// Unlike `FormatDescriptor`/`AggregationDescriptor` there is no legacy
// `label: string` mirror and no `description` — formations have no
// English-only consumer, and the card copy lives with the setup UI.

/** Per-language display names — same convention as `FormatLabels`. Formation
 * names are used as-is in Swedish golf, so `sv` differs only in inflection. */
export interface FormationLabels {
    en: string;
    sv?: string;
}

/** Team size bounds, inclusive. Named (rather than inlined) so the generated
 * clients get a `FormationSize` instead of borrowing a structurally identical
 * type from an unrelated descriptor. */
export interface FormationSize {
    min: number;
    max: number;
}

/**
 * Serializable formation metadata. Carries NO functions —
 * `JSON.parse(JSON.stringify(descriptor))` must round-trip identically (the
 * proposal's `allowances(memberCount)` ships as the per-size table below).
 */
export interface FormationDescriptor {
    /** Stable, language-independent key; matches `DraftRoundTeam.formation`. */
    id: string;
    labels: FormationLabels;
    /** Foursomes/greensomes are exactly 2; scramble is capped at 8, under the
     * setup UI's `MAX_TEAM_SIZE` of 10. */
    size: FormationSize;
    /**
     * Default allowance % by position, keyed by member count. Members are
     * sorted by playing handicap ASCENDING, so position 1 (index 0) is the
     * lowest handicap. Keys are strings because that is what a JSON object
     * carries; every key lies within `size` and every array's length equals its
     * key (formation-catalog.test.ts holds that shut).
     *
     * Defaults are computed, overrides are sticky: the setup UI re-seeds on a
     * membership/handicap change unless the team was hand-edited.
     */
    allowancesBySize: Record<string, number[]>;
}

// --- The catalog ------------------------------------------------------------
//
// Established WHS/USGA conventions for foursomes (50/50) and greensomes
// (60/40); scramble convention exists up to 4 (35/15, 30/20/10, 25/20/15/10)
// and no further — 5-8 extend it with 5 then a 0 floor. That tail is a
// DOCUMENTED DEFAULT, not a rule, and is freely overridable per member.

const FORMATIONS: FormationDescriptor[] = [
    {
        id: 'foursomes',
        labels: { en: 'Foursomes', sv: 'Foursome' },
        size: { min: 2, max: 2 },
        allowancesBySize: { 2: [50, 50] },
    },
    {
        id: 'greensomes',
        labels: { en: 'Greensomes', sv: 'Greensome' },
        size: { min: 2, max: 2 },
        allowancesBySize: { 2: [60, 40] },
    },
    {
        id: 'scramble',
        labels: { en: 'Scramble', sv: 'Scramble' },
        size: { min: 2, max: 8 },
        allowancesBySize: {
            2: [35, 15],
            3: [30, 20, 10],
            4: [25, 20, 15, 10],
            5: [25, 20, 15, 10, 5],
            6: [25, 20, 15, 10, 5, 0],
            7: [25, 20, 15, 10, 5, 0, 0],
            8: [25, 20, 15, 10, 5, 0, 0, 0],
        },
    },
];

/** Serializable catalog — deterministically ordered by id, like
 * `formatCatalog()` / `aggregationCatalog()`. Chip order is a client concern. */
export function formationCatalog(): FormationDescriptor[] {
    return [...FORMATIONS].sort((a, b) => a.id.localeCompare(b.id));
}
